src/sequential/sequential.cpp

```diff
@@ -0,0 +1,275 @@
+#include "sequential.hpp"
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// HELPER CLASSES
+////////////////////////////////////////////////////////////////////////////////
+
+bool
+SequentialBucket::is_empty() {
+  LOG_TRACE("Enter");
+  return this->offset == SIZE_MAX;
+}
+
+void
+SequentialBucket::set(const KeyType key,
+                      const ValueType value,
+                      const HashCodeType hashcode,
+                      const size_t offset) {
+  LOG_TRACE("Enter");
+  assert(this->is_empty() && "overwriting non-empty SequentialBucket");
```

Comment:
is the message just writing to std out? or is it actually checking a condition? sorry never seen an assert like this

Reply:
Good question. The `assert()` function (iirc it's actually a macro) ensures its argument is true. If so, it continues with the operation. Otherwise, it will cause the program to exit and print its line to the screen (I assume it's stderr, but I'm not sure). So the string is just printed to the screen if the assertion is false. It's just a hack tbh. 

N.B. it only exists in debug mode, which is why it's not my favourite.

Reply:
Yeah, @griffinandrew the `&& "some descriptive message"` thing is just a hack to attach messages to the assertion failure that gets printed to `stderr`. Since a string is a pointer, and this pointer is clearly not null, it evaluates to `true`, which is why it works.

Imo it's slightly more readable to put the message first but that's just me being super picky lol.

Reply:
Yeah, I just inherited the convention from here (probably): https://stackoverflow.com/questions/5867834/assert-with-message